// @tanstack/query-core/build/lib/focusManager.esm.js
import { classPrivateFieldLooseBase as _classPrivateFieldLooseBase, classPrivateFieldLooseKey as _classPrivateFieldLooseKey } from './_virtual/_rollupPluginBabelHelpers.esm.js';
import { Subscribable } from './subscribable.esm.js';
import { isServer } from './utils.esm.js';

var _focused = /*#__PURE__*/_classPrivateFieldLooseKey("focused");
var _cleanup = /*#__PURE__*/_classPrivateFieldLooseKey("cleanup");
var _setup = /*#__PURE__*/_classPrivateFieldLooseKey("setup");
class FocusManager extends Subscribable {
  constructor() {
    super();
    Object.defineProperty(this, _focused, {
      writable: true,
      value: void 0
    });
    Object.defineProperty(this, _cleanup, {
      writable: true,
      value: void 0
    });
    Object.defineProperty(this, _setup, {
      writable: true,
      value: void 0
    });
    _classPrivateFieldLooseBase(this, _setup)[_setup] = onFocus => {
      // addEventListener does not exist in React Native, but window does
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      if (!isServer && window.addEventListener) {
        const listener = () => onFocus();
        // Listen to visibilitychange
        window.addEventListener('visibilitychange', listener, false);
        return () => {
          // Be sure to unsubscribe if a new handler is set
          window.removeEventListener('visibilitychange', listener);
        };
      }
      return;
    };
  }
  onSubscribe() {
    if (!_classPrivateFieldLooseBase(this, _cleanup)[_cleanup]) {
      this.setEventListener(_classPrivateFieldLooseBase(this, _setup)[_setup]);
    }
  }
  onUnsubscribe() {
    if (!this.hasListeners()) {
      var _classPrivateFieldLoo, _classPrivateFieldLoo2;
      (_classPrivateFieldLoo = (_classPrivateFieldLoo2 = _classPrivateFieldLooseBase(this, _cleanup))[_cleanup]) == null ? void 0 : _classPrivateFieldLoo.call(_classPrivateFieldLoo2);
      _classPrivateFieldLooseBase(this, _cleanup)[_cleanup] = undefined;
    }
  }
  setEventListener(setup) {
    var _classPrivateFieldLoo3, _classPrivateFieldLoo4;
    _classPrivateFieldLooseBase(this, _setup)[_setup] = setup;
    (_classPrivateFieldLoo3 = (_classPrivateFieldLoo4 = _classPrivateFieldLooseBase(this, _cleanup))[_cleanup]) == null ? void 0 : _classPrivateFieldLoo3.call(_classPrivateFieldLoo4);
    _classPrivateFieldLooseBase(this, _cleanup)[_cleanup] = setup(focused => {
      if (typeof focused === 'boolean') {
        this.setFocused(focused);
      } else {
        this.onFocus();
      }
    });
  }
  setFocused(focused) {
    _classPrivateFieldLooseBase(this, _focused)[_focused] = focused;
    if (focused) {
      this.onFocus();
    }
  }
  onFocus() {
    this.listeners.forEach(listener => {
      listener();
    });
  }
  isFocused() {
    var _globalThis$document;
    if (typeof _classPrivateFieldLooseBase(this, _focused)[_focused] === 'boolean') {
      return _classPrivateFieldLooseBase(this, _focused)[_focused];
    }

    // document global can be unavailable in react native
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    return ((_globalThis$document = globalThis.document) == null ? void 0 : _globalThis$document.visibilityState) !== 'hidden';
  }
}
const focusManager = new FocusManager();

export { FocusManager, focusManager };
