// @tanstack/query-core/build/lib/index.js
'use strict';

var retryer = require('./retryer.js');
var queryCache = require('./queryCache.js');
var queryClient = require('./queryClient.js');
var queryObserver = require('./queryObserver.js');
var queriesObserver = require('./queriesObserver.js');
var infiniteQueryObserver = require('./infiniteQueryObserver.js');
var mutationCache = require('./mutationCache.js');
var mutationObserver = require('./mutationObserver.js');
var notifyManager = require('./notifyManager.js');
var focusManager = require('./focusManager.js');
var onlineManager = require('./onlineManager.js');
var utils = require('./utils.js');
var hydration = require('./hydration.js');



exports.CancelledError = retryer.CancelledError;
exports.isCancelledError = retryer.isCancelledError;
exports.QueryCache = queryCache.QueryCache;
exports.QueryClient = queryClient.QueryClient;
exports.QueryObserver = queryObserver.QueryObserver;
exports.QueriesObserver = queriesObserver.QueriesObserver;
exports.InfiniteQueryObserver = infiniteQueryObserver.InfiniteQueryObserver;
exports.MutationCache = mutationCache.MutationCache;
exports.MutationObserver = mutationObserver.MutationObserver;
exports.notifyManager = notifyManager.notifyManager;
exports.focusManager = focusManager.focusManager;
exports.onlineManager = onlineManager.onlineManager;
exports.hashKey = utils.hashKey;
exports.isServer = utils.isServer;
exports.keepPreviousData = utils.keepPreviousData;
exports.matchQuery = utils.matchQuery;
exports.replaceEqualDeep = utils.replaceEqualDeep;
exports.defaultShouldDehydrateMutation = hydration.defaultShouldDehydrateMutation;
exports.defaultShouldDehydrateQuery = hydration.defaultShouldDehydrateQuery;
exports.dehydrate = hydration.dehydrate;
exports.hydrate = hydration.hydrate;
