// @tanstack/query-core/build/lib/mutationObserver.esm.js
import { classPrivateFieldLooseBase as _classPrivateFieldLooseBase, classPrivateFieldLooseKey as _classPrivateFieldLooseKey } from './_virtual/_rollupPluginBabelHelpers.esm.js';
import { getDefaultState } from './mutation.esm.js';
import { notifyManager } from './notifyManager.esm.js';
import { Subscribable } from './subscribable.esm.js';
import { shallowEqualObjects } from './utils.esm.js';

// TYPES
var _client = /*#__PURE__*/_classPrivateFieldLooseKey("client");
var _currentResult = /*#__PURE__*/_classPrivateFieldLooseKey("currentResult");
var _currentMutation = /*#__PURE__*/_classPrivateFieldLooseKey("currentMutation");
var _mutateOptions = /*#__PURE__*/_classPrivateFieldLooseKey("mutateOptions");
var _updateResult = /*#__PURE__*/_classPrivateFieldLooseKey("updateResult");
var _notify = /*#__PURE__*/_classPrivateFieldLooseKey("notify");
// CLASS

class MutationObserver extends Subscribable {
  constructor(client, options) {
    super();
    Object.defineProperty(this, _notify, {
      value: _notify2
    });
    Object.defineProperty(this, _updateResult, {
      value: _updateResult2
    });
    Object.defineProperty(this, _client, {
      writable: true,
      value: void 0
    });
    Object.defineProperty(this, _currentResult, {
      writable: true,
      value: undefined
    });
    Object.defineProperty(this, _currentMutation, {
      writable: true,
      value: void 0
    });
    Object.defineProperty(this, _mutateOptions, {
      writable: true,
      value: void 0
    });
    _classPrivateFieldLooseBase(this, _client)[_client] = client;
    this.setOptions(options);
    this.bindMethods();
    _classPrivateFieldLooseBase(this, _updateResult)[_updateResult]();
  }
  bindMethods() {
    this.mutate = this.mutate.bind(this);
    this.reset = this.reset.bind(this);
  }
  setOptions(options) {
    var _classPrivateFieldLoo;
    const prevOptions = this.options;
    this.options = _classPrivateFieldLooseBase(this, _client)[_client].defaultMutationOptions(options);
    if (!shallowEqualObjects(prevOptions, this.options)) {
      _classPrivateFieldLooseBase(this, _client)[_client].getMutationCache().notify({
        type: 'observerOptionsUpdated',
        mutation: _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation],
        observer: this
      });
    }
    (_classPrivateFieldLoo = _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation]) == null ? void 0 : _classPrivateFieldLoo.setOptions(this.options);
  }
  onUnsubscribe() {
    if (!this.hasListeners()) {
      var _classPrivateFieldLoo2;
      (_classPrivateFieldLoo2 = _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation]) == null ? void 0 : _classPrivateFieldLoo2.removeObserver(this);
    }
  }
  onMutationUpdate(action) {
    _classPrivateFieldLooseBase(this, _updateResult)[_updateResult]();
    _classPrivateFieldLooseBase(this, _notify)[_notify](action);
  }
  getCurrentResult() {
    return _classPrivateFieldLooseBase(this, _currentResult)[_currentResult];
  }
  reset() {
    _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation] = undefined;
    _classPrivateFieldLooseBase(this, _updateResult)[_updateResult]();
    _classPrivateFieldLooseBase(this, _notify)[_notify]();
  }
  mutate(variables, options) {
    var _classPrivateFieldLoo3;
    _classPrivateFieldLooseBase(this, _mutateOptions)[_mutateOptions] = options;
    (_classPrivateFieldLoo3 = _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation]) == null ? void 0 : _classPrivateFieldLoo3.removeObserver(this);
    _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation] = _classPrivateFieldLooseBase(this, _client)[_client].getMutationCache().build(_classPrivateFieldLooseBase(this, _client)[_client], this.options);
    _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation].addObserver(this);
    return _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation].execute(variables);
  }
}
function _updateResult2() {
  var _classPrivateFieldLoo4, _classPrivateFieldLoo5;
  const state = (_classPrivateFieldLoo4 = (_classPrivateFieldLoo5 = _classPrivateFieldLooseBase(this, _currentMutation)[_currentMutation]) == null ? void 0 : _classPrivateFieldLoo5.state) != null ? _classPrivateFieldLoo4 : getDefaultState();
  _classPrivateFieldLooseBase(this, _currentResult)[_currentResult] = {
    ...state,
    isPending: state.status === 'pending',
    isSuccess: state.status === 'success',
    isError: state.status === 'error',
    isIdle: state.status === 'idle',
    mutate: this.mutate,
    reset: this.reset
  };
}
function _notify2(action) {
  notifyManager.batch(() => {
    // First trigger the mutate callbacks
    if (_classPrivateFieldLooseBase(this, _mutateOptions)[_mutateOptions] && this.hasListeners()) {
      if ((action == null ? void 0 : action.type) === 'success') {
        var _classPrivateFieldLoo6, _classPrivateFieldLoo7, _classPrivateFieldLoo8, _classPrivateFieldLoo9;
        (_classPrivateFieldLoo6 = (_classPrivateFieldLoo7 = _classPrivateFieldLooseBase(this, _mutateOptions)[_mutateOptions]).onSuccess) == null ? void 0 : _classPrivateFieldLoo6.call(_classPrivateFieldLoo7, action.data, _classPrivateFieldLooseBase(this, _currentResult)[_currentResult].variables, _classPrivateFieldLooseBase(this, _currentResult)[_currentResult].context);
        (_classPrivateFieldLoo8 = (_classPrivateFieldLoo9 = _classPrivateFieldLooseBase(this, _mutateOptions)[_mutateOptions]).onSettled) == null ? void 0 : _classPrivateFieldLoo8.call(_classPrivateFieldLoo9, action.data, null, _classPrivateFieldLooseBase(this, _currentResult)[_currentResult].variables, _classPrivateFieldLooseBase(this, _currentResult)[_currentResult].context);
      } else if ((action == null ? void 0 : action.type) === 'error') {
        var _classPrivateFieldLoo10, _classPrivateFieldLoo11, _classPrivateFieldLoo12, _classPrivateFieldLoo13;
        (_classPrivateFieldLoo10 = (_classPrivateFieldLoo11 = _classPrivateFieldLooseBase(this, _mutateOptions)[_mutateOptions]).onError) == null ? void 0 : _classPrivateFieldLoo10.call(_classPrivateFieldLoo11, action.error, _classPrivateFieldLooseBase(this, _currentResult)[_currentResult].variables, _classPrivateFieldLooseBase(this, _currentResult)[_currentResult].context);
        (_classPrivateFieldLoo12 = (_classPrivateFieldLoo13 = _classPrivateFieldLooseBase(this, _mutateOptions)[_mutateOptions]).onSettled) == null ? void 0 : _classPrivateFieldLoo12.call(_classPrivateFieldLoo13, undefined, action.error, _classPrivateFieldLooseBase(this, _currentResult)[_currentResult].variables, _classPrivateFieldLooseBase(this, _currentResult)[_currentResult].context);
      }
    }

    // Then trigger the listeners
    this.listeners.forEach(listener => {
      listener(_classPrivateFieldLooseBase(this, _currentResult)[_currentResult]);
    });
  });
}

export { MutationObserver };
