// @tanstack/query-core/build/lib/queryCache.esm.js
import { classPrivateFieldLooseBase as _classPrivateFieldLooseBase, classPrivateFieldLooseKey as _classPrivateFieldLooseKey } from './_virtual/_rollupPluginBabelHelpers.esm.js';
import { hashQueryKeyByOptions, matchQuery } from './utils.esm.js';
import { Query } from './query.esm.js';
import { notifyManager } from './notifyManager.esm.js';
import { Subscribable } from './subscribable.esm.js';

var _queries = /*#__PURE__*/_classPrivateFieldLooseKey("queries");
// CLASS

class QueryCache extends Subscribable {
  constructor(config = {}) {
    super();
    Object.defineProperty(this, _queries, {
      writable: true,
      value: void 0
    });
    this.config = config;
    _classPrivateFieldLooseBase(this, _queries)[_queries] = new Map();
  }
  build(client, options, state) {
    var _options$queryHash;
    const queryKey = options.queryKey;
    const queryHash = (_options$queryHash = options.queryHash) != null ? _options$queryHash : hashQueryKeyByOptions(queryKey, options);
    let query = this.get(queryHash);
    if (!query) {
      query = new Query({
        cache: this,
        queryKey,
        queryHash,
        options: client.defaultQueryOptions(options),
        state,
        defaultOptions: client.getQueryDefaults(queryKey)
      });
      this.add(query);
    }
    return query;
  }
  add(query) {
    if (!_classPrivateFieldLooseBase(this, _queries)[_queries].has(query.queryHash)) {
      _classPrivateFieldLooseBase(this, _queries)[_queries].set(query.queryHash, query);
      this.notify({
        type: 'added',
        query
      });
    }
  }
  remove(query) {
    const queryInMap = _classPrivateFieldLooseBase(this, _queries)[_queries].get(query.queryHash);
    if (queryInMap) {
      query.destroy();
      if (queryInMap === query) {
        _classPrivateFieldLooseBase(this, _queries)[_queries].delete(query.queryHash);
      }
      this.notify({
        type: 'removed',
        query
      });
    }
  }
  clear() {
    notifyManager.batch(() => {
      this.getAll().forEach(query => {
        this.remove(query);
      });
    });
  }
  get(queryHash) {
    return _classPrivateFieldLooseBase(this, _queries)[_queries].get(queryHash);
  }
  getAll() {
    return [..._classPrivateFieldLooseBase(this, _queries)[_queries].values()];
  }
  find(filters) {
    const defaultedFilters = {
      exact: true,
      ...filters
    };
    return this.getAll().find(query => matchQuery(defaultedFilters, query));
  }
  findAll(filters = {}) {
    const queries = this.getAll();
    return Object.keys(filters).length > 0 ? queries.filter(query => matchQuery(filters, query)) : queries;
  }
  notify(event) {
    notifyManager.batch(() => {
      this.listeners.forEach(listener => {
        listener(event);
      });
    });
  }
  onFocus() {
    notifyManager.batch(() => {
      this.getAll().forEach(query => {
        query.onFocus();
      });
    });
  }
  onOnline() {
    notifyManager.batch(() => {
      this.getAll().forEach(query => {
        query.onOnline();
      });
    });
  }
}

export { QueryCache };
